File: python/2127.py
from typing import List

class Solution:
    def maximumInvitations(self, favorite: List[int]) -> int:
        n = len(favorite)
        deg = [0] * n
        max_depth = [1] * n
        for f in favorite:
            deg[f] += 1
        q = []
        for i, d in enumerate(deg):
            if d == 0: q.append(i)
        while q:
            node = q.pop()
            nxt = favorite[node]
            deg[nxt] -= 1 
            max_depth[nxt] = max(max_depth[node] + 1, max_depth[nxt])
            if deg[nxt] == 0: q.append(nxt)
        max_ring_size = 0
        max_line_size = 0
        for i, d in enumerate(deg):
            if d != 0:
                deg[i] = 0
                nxt = favorite[i]
                ring_size = 1
                while nxt != i:
                    ring_size += 1
                    deg[nxt] = 0
                    nxt = favorite[nxt]
                if ring_size == 2:
                    max_line_size += max_depth[i] + max_depth[favorite[i]] 
                else:
                    max_ring_size = max(max_ring_size, ring_size)
                print(max_ring_size, max_line_size, i, favorite[i])
        return max(max_ring_size, max_line_size)

File: go/2127.go
package main

func maximumInvitations(favorite []int) int {
	n := len(favorite)
	deg := make([]int, n)
	maxDepth := make([]int, n)
	for _, f := range favorite {
		deg[f]++
	}
	q := []int{}
	for i, d := range deg {
		if d == 0 {
			q = append(q, i)
		}
	}
	for len(q) > 0 {
		node := q[0]
		q = q[1:]
		nxt := favorite[node]
		maxDepth[nxt] = maxDepth[node] + 1
		deg[nxt]--
		if deg[nxt] == 0 {
			q = append(q, nxt)
		}
	}
	maxRingSize := 0
	maxLineSize := 0
	for i, d := range deg {
		if d != 0 {
			deg[i] = 0
			ringSize := 1
			x := favorite[i]
			for x != i {
				ringSize++
				deg[x] = 0
				x = favorite[x]
			}
			if ringSize == 2 {
				maxLineSize += maxDepth[i] + maxDepth[favorite[i]] + 2
			} else {
				maxRingSize = max(maxRingSize, ringSize)
			}
		}
	}
	return max(maxLineSize, maxRingSize)
}
